src/slice/slice_arc.rs

```diff
@@ -0,0 +1,60 @@
+// Copyright (c) 2024-present, fjall-rs
+// This source code is licensed under both the Apache 2.0 and MIT License
+// (found in the LICENSE-* files in the repository)
+
+use std::sync::Arc;
+
+/// An immutable byte slice that can be cloned without additional heap allocation
+#[derive(Debug, Clone, Eq, Hash, Ord)]
+pub struct Slice(Arc<[u8]>);
+
+impl Slice {
+    /// Construct a [`Slice`] from a byte slice.
+    #[must_use]
+    pub fn new(bytes: &[u8]) -> Self {
+        Self(bytes.into())
+    }
+
+    #[must_use]
+    #[doc(hidden)]
+    pub fn with_size(len: usize) -> Self {
+        // TODO: optimize this with byteview to remove the reallocation
+        let v = vec![0; len];
+        Self(v.into())
+    }
+
+    #[doc(hidden)]
+    pub fn from_reader<R: std::io::Read>(reader: &mut R, len: usize) -> std::io::Result<Self> {
+        let mut view = Self::with_size(len);
+        let builder = Arc::get_mut(&mut view.0).expect("we are the owner");
+        reader.read_exact(builder)?;
+        Ok(view)
+    }
+}
+
+impl AsRef<[u8]> for Slice {
```

Comment:
The AsRefs can probably be moved too, or does that not compile?

Reply:
I could if I make `self.0` public to the parent module. I generally like to avoid that, but up to you.